core/shared/src/main/scala/org/http4s/Service.scala

```diff
@@ -0,0 +1,128 @@
+      }
+    }
+
+  def orElse(default: => Service[F, A, B]): Service[F, A, B] =
+    new Service[F, A, B] {
+      def applyOrElse(a: A, default2: A => Resource[F, B]): Resource[F, B] = {
+        val rb = self.applyOrElse(a, checkFallback)
+        if (isFallback(rb)) default.applyOrElse(a, default2)
+        else rb
+      }
+    }
+
+  def local[C](f: C => A): Service[F, C, B] =
+    new Service[F, C, B] {
+      def applyOrElse(c: C, default: C => Resource[F, B]): Resource[F, B] = {
+        val rb = self.applyOrElse(f(c), checkFallback)
+        if (isFallback(rb)) default(c) else rb
+      }
+    }
+}
+
+object Service extends ServiceInstances {
+  // This is the same basic trick as used in scala.PartialFunction to
+  // detect partiality without allocating an option.  We use
+  // fallbackSentinel as the default function in calls to applyOrElse.
+  // If the call returns the fallbackSentinel, it fell through and we
+  // need to invoke our own default logic.
+  private[this] val fallbackSentinel: Any => AnyRef = Function.const(fallbackSentinel)
+  private def checkFallback[F[_], B] = fallbackSentinel.asInstanceOf[Any => Resource[F, B]]
+  private def isFallback[F[_], B](x: F[B]) = fallbackSentinel eq x.asInstanceOf[AnyRef]
+
+  def pure[F[_], A, B](b: B): Service[F, A, B] =
+    new Service[F, A, B] {
+      val rb = Resource.pure[F, B](b)
+      def applyOrElse(a: A, default: A => Resource[F, B]): Resource[F, B] =
+        rb
+    }
+
+  def pass[F[_], A, B]: Service[F, A, B] =
+    new Service[F, A, B] {
+      def applyOrElse(a: A, default: A => Resource[F, B]): Resource[F, B] =
+        default(a)
+    }
+
+  def ask[F[_], A]: Service[F, A, A] =
+    new Service[F, A, A] {
+      def applyOrElse(a: A, default: A => Resource[F, A]): Resource[F, A] =
+        Resource.pure(a)
+    }
+}
+
+private[http4s] sealed abstract class ServiceInstance[F[_], A]
+    extends Monad[Service[F, A, *]]
+    with StackSafeMonad[Service[F, A, *]]
+    with Alternative[Service[F, A, *]] {
+  override def map[B, C](fb: Service[F, A, B])(f: B => C): Service[F, A, C] =
+    fb.map(f)
+
+  def pure[B](b: B): Service[F, A, B] =
+    Service.pure(b)
+
+  def flatMap[B, C](fb: Service[F, A, B])(f: B => Service[F, A, C]): Service[F, A, C] =
+    fb.flatMap(f)
+
+  def combineK[B](x: Service[F, A, B], y: Service[F, A, B]): Service[F, A, B] =
+    x.orElse(y)
+
+  def empty[B]: Service[F, A, B] =
+    Service.pass
+}
+
+private[http4s] trait ServiceInstances {
+  implicit def catsInstancesForHttp4sService[F[_], A]
+      : Monad[Service[F, A, *]] with Alternative[Service[F, A, *]] =
+    new ServiceInstance[F, A] {}
+
+  implicit def catsMtlLocalForHttp4sService[F[_], A]: Local[Service[F, A, *], A] =
+    new Local[Service[F, A, *], A] {
+      val applicative = Applicative[Service[F, A, *]]
+      def ask[E >: A]: Service[F, A, E] =
+        Service.ask.asInstanceOf[Service[F, A, E]]
```

Comment:
Shady.  I got rid of the covariance of A.  I can try to bring it back, but it has to be on applyOrElse, and that ripples a lot of other places.